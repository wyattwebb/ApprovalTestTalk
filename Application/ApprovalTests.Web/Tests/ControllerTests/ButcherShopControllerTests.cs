using System.Collections.Generic;
using System.Linq;
using ApprovalTests.Reporters;
using ApprovalTests.Web.Controllers;
using ApprovalTests.Web.Models.BaconViewModels;
using ApprovalTests.Web.PersistanceModels.BaconModels;
using ApprovalTests.Web.Services;
using ApprovalTests.Web.Tests.Setup;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using Moq;
using Ploeh.AutoFixture;

namespace ApprovalTests.Web.Tests.ControllerTests
{
    [TestClass]
    public class ButcherShopControllerTests : ApiTestBase
    {

        private readonly Fixture _fixture;

        public ButcherShopControllerTests()
        {
            _fixture = new Fixture();
            _fixture.Customize(new ApprovalTestCustomization());
        }

        protected override string BaseUrl
        {
            get { return "http://localhost:8000/api/"; }
        }

        [TestMethod]
        public void GET_GetPigs__Given_10Pigs__When_Mapper_Succeeds__Then_Expected_Mapped_Result()
        {
            // Arrange
            var count = 10;
            var pigs = _fixture.CreateMany<Pig>(10).ToArray();

            var turkeyBaconService = new Mock<IBaconService>();
            turkeyBaconService.Setup(x => x.GetGeneratedPigs(count))
                .Returns(() => pigs);

            var subject = new ButcherShopController(
                baconService: turkeyBaconService.Object);

            // Act
            var result = subject.GetPigs(count);

            // Assert
            Assert.AreEqual(pigs[0].Belly, result.Pigs[0].Belly);
            Assert.AreEqual(pigs[0].Back, result.Pigs[0].Back);
            Assert.AreEqual(pigs[0].Cannon, result.Pigs[0].Cannon);
            Assert.AreEqual(pigs[0].Cheek, result.Pigs[0].Cheek);
            Assert.AreEqual(pigs[0].Coffin, result.Pigs[0].Coffin);
            Assert.AreEqual(pigs[0].Crops, result.Pigs[0].Crops);
            Assert.AreEqual(pigs[0].Dewclaw, result.Pigs[0].Dewclaw);
            Assert.AreEqual(pigs[0].Ear, result.Pigs[0].Ear);
            // ... Assert the rest
            Assert.AreEqual(pigs[1].Belly, result.Pigs[0].Belly);
            Assert.AreEqual(pigs[1].Back, result.Pigs[0].Back);
            Assert.AreEqual(pigs[1].Cannon, result.Pigs[0].Cannon);
            Assert.AreEqual(pigs[1].Cheek, result.Pigs[0].Cheek);
            Assert.AreEqual(pigs[1].Coffin, result.Pigs[0].Coffin);
            Assert.AreEqual(pigs[1].Crops, result.Pigs[0].Crops);
            Assert.AreEqual(pigs[1].Dewclaw, result.Pigs[0].Dewclaw);
            Assert.AreEqual(pigs[1].Ear, result.Pigs[0].Ear);
            // ... Assert the rest up till 10

        }

        [TestMethod, Description("Approval Tests Assert")]
        public void GET_GetPigs_Ten()
        {
            // Arrange
            var count = 10;
            var pigs = _fixture.CreateMany<Pig>(10).ToArray();

            var turkeyBaconService = new Mock<IBaconService>();
            turkeyBaconService.Setup(x => x.GetGeneratedPigs(count))
                .Returns(() => pigs);

            var subject = new ButcherShopController(
                baconService: turkeyBaconService.Object);
            
            // Act
            var result = subject.GetPigs(count);

            // Assert
            VerifyApprovedJsonResult(result);
        }

        [TestMethod, Description("No AutoFixture")]
        public void GET_Given_Manual_Setups__When_GetPigs_Three__ThenAsserts()
        {
            // Arrange
            var count = 3;
            var pigs = new[]
            {
                new Pig
                {
                    Back = "Back",
                    Belly = "Belly",
                    Cannon = "Cannon",
                    Cheek = "Cheek",
                    Coffin = "Coffin",
                    Crops = "Crops",
                    Dewclaw = "Dewclaw",
                    Ear = "Ear",
                    FetLock = "FetLock",
                    ForeFlank = "ForFlank"
                    // TODO the rest..
                },
                new Pig
                {
                    Back = "Back",
                    Belly = "Belly",
                    Cannon = "Cannon",
                    Cheek = "Cheek",
                    Coffin = "Coffin",
                    Crops = "Crops",
                    Dewclaw = "Dewclaw",
                    Ear = "Ear",
                    FetLock = "FetLock",
                    ForeFlank = "ForFlank"
                    // TODO the rest..
                },
                new Pig
                {
                    Back = "Back",
                    Belly = "Belly",
                    Cannon = "Cannon",
                    Cheek = "Cheek",
                    Coffin = "Coffin",
                    Crops = "Crops",
                    Dewclaw = "Dewclaw",
                    Ear = "Ear",
                    FetLock = "FetLock",
                    ForeFlank = "ForFlank"
                    // TODO the rest..
                }
            };

            var turkeyBaconService = new Mock<IBaconService>();
            turkeyBaconService.Setup(x => x.GetGeneratedPigs(count))
                .Returns(() => pigs);

            var subject = new ButcherShopController(
                baconService: turkeyBaconService.Object);

            // Act
            var result = subject.GetPigs(count);

            // Assert
            VerifyApprovedJsonResult(result);
        }

        [TestMethod, Description("Approval Tests Assert With AutoFixture")]
        public void GET_GetPigs_Ten_UsingAutoFixture()
        {
            // Arrange
            var count = 10;
            var pigs = _fixture.CreateMany<Pig>(10).ToArray();

            var turkeyBaconService = new Mock<IBaconService>();
            turkeyBaconService.Setup(x => x.GetGeneratedPigs(count))
                .Returns(() => pigs);

            var subject = new ButcherShopController(
                baconService: turkeyBaconService.Object);

            // Act
            var result = subject.GetPigs(count);

            // Assert
            VerifyApprovedJsonResult(result);
        }

        [TestMethod, Description("Manual Asserts on null Validation")]
        public void GET_GetPigs_Given_Null_Parameter__When_Executed__Then_Return_Validation_Error_ManualAssertion()
        {
            // Arrange
            int? count = null;

            var expected = "Count is Required";
            var mockValidation = new Mock<IValidateInput>();
            mockValidation.Setup(x => x.ValidateGetPigs(count))
                .Returns(expected);

            var subject = new ButcherShopController(
                validateInput: mockValidation.Object);

            // Act
            var result = subject.GetPigs(count);

            // Assert
            Assert.AreEqual(result.ErrorMessage, expected);
        }

        [TestMethod, Description("Manual Asserts on min Validation")]
        public void GET_GetPigs_Given_Min_Parameter__When_Executed__Then_Return_Validation_Error_ManualAssertion()
        {
            // Arrange
            int? count = 5;

            var expected = "FIVE is OUTRIGHT!";
            var mockValidation = new Mock<IValidateInput>();
            mockValidation.Setup(x => x.ValidateGetPigs(count))
                .Returns(expected);

            var subject = new ButcherShopController(
                validateInput: mockValidation.Object);

            // Act
            var result = subject.GetPigs(count);

            // Assert
            Assert.AreEqual(result.ErrorMessage, expected);
        }

        [TestMethod, Description("Multiple Approval Asserts")]
        public void GET_GetPigs_Given_Parameter__When_Executed__Then_Return_Validation_Errors()
        {
            // Arrange
            int? nullCount = null;
            var fiveCount = 5;

            var nullExpected = "Count is Required";
            var mockValidation = new Mock<IValidateInput>();
            mockValidation.Setup(x => x.ValidateGetPigs(nullCount))
                .Returns(nullExpected);

            var fiveExpected = "FIVE is OUTRIGHT!";
            mockValidation.Setup(x => x.ValidateGetPigs(fiveCount))
                .Returns(fiveExpected);

            var subject = new ButcherShopController(
                validateInput: mockValidation.Object);
            var results = new List<PigsViewModel>();

            // Act
            results.Add(subject.GetPigs(nullCount));
            results.Add(subject.GetPigs(fiveCount));

            // Assert
            VerifyApprovedJsonResult(results);
        }

        [TestMethod, Description("Multiple Approval Asserts HTML")]
        public void GET_GetPigs_Given_Parameter__When_Executed__Then_Return_Validation_Errors_Html()
        {
            // Arrange
            int? nullCount = null;
            var fiveCount = 5;

            var nullExpected = "Count is Required";
            var mockValidation = new Mock<IValidateInput>();
            mockValidation.Setup(x => x.ValidateGetPigs(nullCount))
                .Returns(nullExpected);

            var fiveExpected = "FIVE is OUTRIGHT!";
            mockValidation.Setup(x => x.ValidateGetPigs(fiveCount))
                .Returns(fiveExpected);

            var subject = new ButcherShopController(
                validateInput: mockValidation.Object);
            var results = new List<PigsViewModel>();

            // Act
            var htmlReport = new BuildHtml("Multiple Approval Test Count Input");
            htmlReport.AddLine(subject.GetPigs(nullCount).ErrorMessage);
            htmlReport.AddLine(subject.GetPigs(fiveCount).ErrorMessage);

            // Assert
            VerifyApprovedHtmlResult(htmlReport.GetHtml());
        }

        [TestMethod]
        [UseReporter(typeof(DiffReporter))]
        public void GET_GetPigs_API_JsonTest()
        {
            var result = ExecuteGetRequest<PigsViewModel>("ButcherShop/GetPigs?id=1");

            VerifyApprovedJsonResult(result);
        }

    }
}
