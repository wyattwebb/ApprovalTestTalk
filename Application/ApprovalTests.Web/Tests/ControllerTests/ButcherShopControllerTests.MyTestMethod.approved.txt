{
  "ErrorMessage": null,
  "Pigs": [
    {
      "Back": "Back4440d833-6190-436f-b7d0-99fcba86e64c",
      "Belly": "Belly2f6e6d86-1a87-41d9-8a9f-487d6190e138",
      "Cannon": "Cannone237a656-c160-406f-86f2-61456974161d",
      "Cheek": "Cheek82b216bb-7e00-4a7f-a5e6-805e9b9e2768",
      "Coffin": "Coffin71c7df59-bbad-4266-922d-81d2e76eda69",
      "Crops": "Crops64de9764-01cb-4dd6-ab99-04abdb33d4fd",
      "Dewclaw": "Dewclawe4603d49-191a-4694-9f66-923a87d2d44f",
      "Ear": "Ear27dd1a4e-d514-43d1-ae54-6bfacab8355e",
      "FetLock": "FetLock3d154e74-f7aa-4057-8f09-92c02b080d72",
      "ForeFlank": "ForeFlank55a5d724-1404-4df7-86da-669fbdcaec15",
      "ForeLeg": "ForeLegcf7b4334-d2ab-43ed-bea2-8ddbc047baec",
      "Ham": "Hama9a72295-667d-485c-b479-277a247e8a4d",
      "Hock": "Hock9a23d848-68dc-4d7a-86b6-405af3207911",
      "Jowls": "Jowls867bec2a-a1de-4317-b35f-60716bb9933a",
      "Knee": "Kneebcbca96f-342b-405b-bada-ea9cec40a8ab",
      "Loin": "Loine4c8e327-29dc-4c26-9dae-c9fa5955e729",
      "Neck": "Neck0ba9d5a2-5087-431b-aded-c132ddd6b833",
      "Nostrils": "Nostrils1469bb04-6820-45eb-91de-2b03047c17fd",
      "Pastern": "Pastern2a3e1550-550d-4bd1-a4d7-f02a705ddb68",
      "Poll": "Poll0e465037-90d7-4b5f-a7c5-d965d865fe5d",
      "RearFlank": "RearFlank6bf88b73-0c00-4fbe-a970-f880f31e94ae",
      "Rump": "Rump1434fbfb-af8e-408f-8df2-fe6fcda689a0",
      "Sheath": "Sheathdc8247a5-bea7-4cd1-be40-d1cd5ff2bfe0",
      "Shoulder": "Shoulder5cb9ca10-97ad-48f9-9b44-284f479d05a4",
      "Snout": "Snoutdf0c21cb-2f4f-4ce8-a289-298a0f6072ee",
      "Stiffle": "Stifflef05740b6-0bbe-4c76-bc02-234aa38d59bd",
      "Tail": "Tail42e413cb-44dc-4f41-b4d8-1071ea4639b2"
    }
  ]
}